README.md
# Linux MinUP

This repository contains the linux files to upload minION data to minoTour.

There are two ways of monitoring minION data using minoTour. The first is by directly processing the fast5 read files containing basecalled minION data. The second is by directly monitoring the activity of the minKNOW software itself directly. These two options can be combined to monitor both reads and minKNOW activity. This helps give remote real time feedback on minION activity. The speed of data analysis is affected by many things, but by enabling direct monitoring of minKNOW, some of these effects can be mitigated.

Brief running instructions are provided below. Further details can be found by watching the tutorial videos available on minoTour.

# Firewalls

MinUP assumes the following ports and connections are open:

port 3306 outgoing to minotour.nottingham.ac.uk

port 80 to minotour.nottingham.ac.uk

port 8080 to minotour.nottingham.ac.uk

# Updates

minUP checks for new versions. We strongly recommend updating to the latest version whenever it is released.

Changes are noted in the 'changes.txt' document.

# Aligning Data

minUP can align reads to a reference in real time. However - you are cautioned not to use this for references larger than 10 Mb at this time. We are working on methods to track real time aligning on larger genomes in the future. At this time, algining to large genomes will significantly slow down the speed of data upload to minoTour.

# Real Time Assumption

minUP and minoTour assume that you are uploading data as they are generated by minKNOW and any of the available basecallers (metrichor, minKNOW, nanonet or albacore). If this is not the case, then the time dependent data sets that are reported in minoTour will only be correct at the end of a run. 

# Running minUP/mincontrol

Files are uploaded using the program minUP.py . An optional GUI wrapper is included - minUPgui.py. 

To run the gui, type:

  `python minUPgui.py`

from the command line in the minUP folder.

This will launch the minUP graphical user interface (GUI). The fields in the GUI will be retained from launch to launch, so once you have configured it, you only need to change individual commands each time you run it. The gui provides a simpler method of running the command line script and provides access to the most common options.

To run the command line script, type:

  `python minUP.py`
  
Help is available by typing:

  `python minUP.py -h`

If all this is too long to read, a simple command for uploading data to minoTour in Nottingham would be:

 `python minUP.py -w /path/to/basecalled/fast5/ -u yourusername -dbu yourusername -pw yourpassword -dbh minotour.nottingham.ac.uk`
 
 If you wish to align to a reference:
 
  `python minUP.py -w /path/to/basecalled/fast5/ -u yourusername -dbu yourusername -pw yourpassword -dbh minotour.nottingham.ac.uk -f /path/to/your/reference.fasta -bwa`

More detailed notes on all the options are available below.

If you are using the GUI, these options are set within the GUI interface itself.

# Real Time Monitoring of minKNOW with minUP

We note that this is a beta function - it is still in development. It may break with minKNOW updates and changes. It may break with lots of users. It may break at any time.

To enable real time monitoring of minKNOW you need to add two extra flags to your minUP command.
These are:

  -ip
  
  -pin
  
-ip is the IP address of the machine running minKNOW. Typically this is 127.0.0.1. Advanced users may wish to run this software on another machine - but note you need extremely permissive firewall settings to do so.
-pin is a security pin number. This is required for any user of minoTour to set automated control features.

So - to run real time monitoring enter (but please don't use the pin 1234!):

  `python minUP.py -w /path/to/basecalled/fast5/ -u yourusername -dbu yourusername -pw yourpassword -dbh minotour.nottingham.ac.uk -ip 127.0.0.1 -pin 1234`
  
  
# Standalone Real Time Monitoring of minKNOW

It is possible to just monitor minKNOW using our software without processing read files. This is done using:

  `python mincontrol.py`
  
No gui interface is available for this option at this time.

A run command for mincontrol.py to connect to minotour.nottingham.ac.uk is (again - please don't use 1234!):

  `python mincontrol.py -ws minotour.nottingham.ac.uk -dbh minotour.nottingham.ac.uk -dbu yourusername -pw yourpassword -pin 1234 -ip 127.0.0.1`


# minUP Command Line Options
The command line options are described below:

  -h, --help            show this help message and exit
  
  -w WATCHDIR, --watch-dir WATCHDIR
  
                        The path to the folder containing the downloads
                        
                        directory with fast5 reads to analyse - e.g.
                        
                        C:\data\minion\downloads (for windows).
                        
  -downloads DOWNLOADS, --downloads-dir DOWNLOADS
  
                        The path to the folder containing the downloads
                        
                        directory with fast5 reads to analyse - e.g.
                        
                        C:\data\minion\downloads (for windows).
                        
  -uploaded UPLOADED, --uploaded-dir UPLOADED
  
                        The path to the folder containing the uploaded
                        
                        directory with fast5 reads to analyse - e.g.
                        
                        C:\data\minion\downloads (for windows).
                        
  -dbh DBHOST, --mysql-host DBHOST
  
                        The location of the MySQL database. default is
                        
                        'localhost'.
                        
  -dbu DBUSERNAME, --mysql-username DBUSERNAME
  
                        The MySQL username with create & write privileges on
                        
                        MinoTour.
                        
  -dbp DBPORT, --mysql-port DBPORT
  
                        The MySQL port number, else the default port '3306' is
                        
                        used.
                        
  -pw DBPASS, --mysql-password DBPASS
  
                        The password for the MySQL username with permission to
                        
                        upload to MinoTour.
                        
  -f REF_FASTA, --align-ref-fasta REF_FASTA
  
                        The reference fasta file to align reads against. Using
                        
                        this option enables read alignment provided LastAl and
                        
                        LastDB are in the path. Leaving this entry blank will
                        
                        upload the data without any alignment. To use multiple
                        
                        reference fasta files input them as one text string
                        
                        seperated by commas (no white spaces)
                        
  -b, --align-batch-fasta
  
                        Align reads in batch processing mode. Assumes the
                        
                        watch-dir (-w) is pointed at a directory with one or
                        
                        more "downloads" folders below it somewhere. Each
                        
                        "downloads" folder can have a subfolder named
                        
                        "reference" containing the fasta file(s) to align the
                        
                        fast5 reads in the corresponding "downloads" folder to
                        
  -procs PROCS, --proc_num PROCS
  
                        The number of processors to run this on.
                        
  -u MINOTOURUSERNAME, --minotour-username MINOTOURUSERNAME
  
                        The MinoTour username with permissions to upload data.
                        
  -s VIEW_USERS, --minotour-sharing-usernames VIEW_USERS
  
                        A comma seperated list (with no whitespaces) of other
                        
                        MinoTour users who will also be able to view the data.
                        
  -o FLOWCELL_OWNER, --flowcell-owner FLOWCELL_OWNER
  
                        The name of the minion owner. 'minionowner' is the
                        
                        default
                        
  -r RUN_NUM, --run-number RUN_NUM
  
                        The run number of the flowcell. The default value is
                        
                        0.
                        
  -c, --commment-true   Add a comment to the comments field for this run.
  
                        Follow the prompt once minup starts .
                        
  -last, --last-align-true
  
                        align reads with LAST
                        
  -bwa, --bwa-align-true
  
                        align reads with BWA
                        
  -bwa-opts BWA_OPTIONS, --bwa-align-options BWA_OPTIONS
  
                        BWA options: Enter a comma-seperated list of BWA
                        
                        options without spaces or '-' characters e.g. k12,T0
                        
  -last-opts LAST_OPTIONS, --last-align-options LAST_OPTIONS
  
                        LAST options: Enter a comma-seperated list of LAST
                        
                        options without spaces or '-' characters e.g.
                        
                        s2,T0,Q0,a1
                        
  -pin PIN, --security-pin PIN
  
                        pin number for remote control
                        
  -ip IP_ADDRESS, --ip-address IP_ADDRESS
  
                        Used for remote control with option '-pin'. Provide IP
                        
                        address of the computer running minKNOW. The default
                        
                        is the IP address of this computer
                        
  -t, --insert-tel-true
  
                        DEPRECATED
                        
  -pre, --insert-pre-true
  
                        Process raw reads prior to basecalling by metrichor.
                        
  -prealign, --align-pre-true
  
                        Align raw reads prior to basecalling by metrichor.
                        
  -d, --drop-db-true    Drop existing database if it already exists.
  
  -v VERBOSE, --verbose-true VERBOSE
  
                        Print detailed messages while processing files.
                        
  -name CUSTOM_NAME, --name-custom CUSTOM_NAME
  
                        Provide a modifier to the database name. This allows
                        
                        you to upload the same dataset to minoTour more than
                        
                        once. The additional string should be as short as
                        
                        possible.
                        
  -cs ADDED_COMMENT [ADDED_COMMENT ...], --commment-string ADDED_COMMENT [ADDED_COMMENT ...]
  

changes.txt
0.75_28.3.17

* Fix to handle metrichor file type changes.

--------------------------------------------------------------------------------

0.74_7.3.17

* Tuning minControl to send less frequent data updates.

--------------------------------------------------------------------------------

0.74_24.02.17

* Bugfix to deal with time stamp changes in minKNOW 1.4.2
* Fix to handle incorrect reporting of non-existent minIONs.
* Fix to script launching to deal with changes in minKNOW 1.4.2
!! To do: minUPgui launching is temperamental on linux

--------------------------------------------------------------------------------

0.73_14.02.17

* Bugfix to minUPgui to prevent hardcoding of username.


--------------------------------------------------------------------------------

0.72_1.02.17

* Bugfix to events in sql tables timing.

--------------------------------------------------------------------------------

0.71_30.01.17

* Bugfix to deal with global time stamps for API.

--------------------------------------------------------------------------------

0.70_29.01.17

* Updated mincontrol to version 0.3

--------------------------------------------------------------------------------

0.70_25.01.17

* modified mincontrol to post data only every 30 seconds.
* Fixed hdfTimes sorting bug on evenets['start'] lookup


--------------------------------------------------------------------------------

0.70_24.01.17

* Fixes for get_main_timings() start_time/sampling_rate in config_general
*

--------------------------------------------------------------------------------

0.70_23.01.17

* Implemented "cursor_execute(args,db,cursor,sql)" function in db.py to replace calls to cursor.execute() -- so as to reconnect to mySQL if connection dropped.
* Wired in relevant function arguments ...(args,dc,cursor) throughout all calls to cursor_execute()
* Attempt to use db class was unsucessful -- probably due to connection-pool complexity.
* Adopted ML NYork startMincontrol.py script -- seems to be largely same as existing one.


--------------------------------------------------------------------------------

0.70_18.01.17

* Added alignerBlockSize options to minUP cmd line and gui
* Added optinal args.downloads and args.uploaded to minUP cmd line
* Added args.downloaded = args.watchdir and args.uploaded = 'uploaded' to minUP.py
* Set changed all instances of 'downloads' and 'uploaded' in MyHandler.py to args.dwwnloads and args.uploaded
--------------------------------------------------------------------------------

0.70_05.01.17

* Fixed gui for "pin" option

--------------------------------------------------------------------------------

0.70_21.12.16

* Wholesale re-structuring of mT_align.py
* Rename mT_align.py to mT_coverage.py
* Created separate module startMincontrol invoked from checkreads line 778
* Moved startMNincontrol invokation to MyHandler line 260


--------------------------------------------------------------------------------
0.68_16.12.16

* Checking with ver7 and fixes for on_moved etc
* New mincontrol
* Fixes for minUPgui on OSX etc.

--------------------------------------------------------------------------------
0.68_23.11.16

* Building mincontrolv4 with autobahn
* Fixed py2exe zope build problems by adding empty __init__.py to site-packages/zope/

--------------------------------------------------------------------------------
0.68_21.11.16

# New mincontrolv3.py

--------------------------------------------------------------------------------
0.68_16.11.16

* Changes for new mincontrolv3.py

--------------------------------------------------------------------------------
0.68_18.10.16

* Incermented version to 0.68

--------------------------------------------------------------------------------
0.67_17.10.16

* Fixes to depricate pre_align tables

--------------------------------------------------------------------------------
0.67_14.10.16

* Fixed terminate minup codes for minUPgui on nix systems
* Fixed minUPgui for nix ctrl-c to minUP

--------------------------------------------------------------------------------
0.67_11.10.16

* Fixed minUPgui for nix
* Fixed munUP ver.txt online checking for offline users
* Deprecated
    -procs
    -n
    -prealign
    -qryStartEnd
    -qScale
    -standalone
    -largeRef
    -telem


--------------------------------------------------------------------------------
0.67_10.10.16

* Incremnented version to 0.67
* Fixes for basename from filenames for intergrated and nanonet basecallingfats5 files.


--------------------------------------------------------------------------------
0.66_24.09.16

* Fixed metricoore basecalled_templat qual scores bug

--------------------------------------------------------------------------------
0.66_24.09.16

* Many bug fixes for handling invalid hdf files
* Move file under debug flag control
* Python stylisitc changes etc.

--------------------------------------------------------------------------------
0.65_26.09.16

* Fixed python script tab-space issues for uploads to gdrive
* Fixed dat and bwa issues for new file types
* Fixed prealign for r7 file types
* Testing on r7 and three types of r9 files.
* Fixed window size of minUPgui

--------------------------------------------------------------------------------
0.65_19.09.16

* Fixes for local base called nanonet and integrated RNN (minknow 1.0) files

--------------------------------------------------------------------------------
0.65_18.08.16

* fixes for metrichor r9 basecalled files


minUP.py
#!/gusr/bin/python
# -*- coding: utf-8 -*-

# --------------------------------------------------
# File Name: minup.v0.63.py
# Purpose: minup: a program to process & upload MinION fast5 files
#               in to the minoTour website in real-time or post-run.
# Creation Date: 2014 - 2016
# Last Modified: Wed, Jan 25, 2017  2:40:49 PM
# Author(s): written & designed by
#               Martin J. Blythe, Fei Sang, Mike Stout & Matt W. Loose
#               The DeepSeq Team, University of Nottingham, UK
# Copyright: 2015 The Author(s) All Rights Reserved
# Credits:
# --------------------------------------------------
import sys
import os
import time
from watchdog.observers.polling import PollingObserver as Observer
import re
import MySQLdb
import configargparse
import multiprocessing
import platform  # MS
import signal

# Unbuffered IO
# NB 1 is essential to catch ctrl-c ...
sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 1)  # MS
sys.stderr = os.fdopen(sys.stderr.fileno(), 'w', 0)  # MS


sys.path.append('modules')
from exit_gracefully import terminate_minup
from processRefFasta import process_ref_fasta
#from telem import *

from sql import okSQLname

import urllib

# ---------------------------------------------------------------------------
def toFloat(s):
    a,b = s[:-1].split('_')
    s_ = '.'.join([a,b])
    a,v,d,m,y = s_.split('.')
    s__ = ''.join([a,v,'.',y,m,d])
    return float(s__)
    
def emergencyexit():
    print 'stopping monitor....'
    observer.stop()
    terminate_minup(args, dbcheckhash, OPER, MINUP_VERSION)
    
def emergencyexit2(signum,frame):
    print 'stopping monitor....'
    observer.stop()
    terminate_minup(args, dbcheckhash, OPER, MINUP_VERSION)

# main
if __name__ == '__main__':

    multiprocessing.freeze_support()  # MS

    manager = multiprocessing.Manager()

    global MINUP_VERSION

    fh = open("ver.txt", "r")
    ver_txt = fh.readline()
    fh.close()

    ver_txt_  = toFloat(ver_txt)

    # Check online for new version ....
    try:
        link = "https://raw.githubusercontent.com/minoTour/linminUP/master/ver.txt"
        f = urllib.urlopen(link)
        current_ver = f.read()
        

        current_ver_ = toFloat(current_ver)

        if ver_txt_ < current_ver_:
            print "You are using minUP version %s\nThe latest version is %s\nPlease download the latest version from MINOTOUR" % (ver_txt, current_ver)
            sys.exit()
    except:
        pass




    MINUP_VERSION = re.sub("[^.0-9]", "", ver_txt.split('_')[0])
    __version__ = MINUP_VERSION

    global OPER

    OPER = platform.system()
    if OPER is 'Windows':  # MS
        OPER = 'windows'
    else:
        OPER = 'linux'  # MS
    print OPER  # MS

    global config_file
    global logfolder
    global VALID_REF_DIR
    global BWA_INDEX_DIR
    global LAST_INDEX_DIR
    xml_file_dict = dict()




                # # linux version
    if OPER is 'linux':
        config_file = os.path.join(os.path.sep, \
                                   os.path.dirname(os.path.realpath('__file__' \
                                   )), 'minup_posix.config')
        logfolder = os.path.join(os.path.sep, \
                                 os.path.dirname(os.path.realpath('__file__' \
                                 )), 'minup_run_logs')
        VALID_REF_DIR = os.path.join(os.path.sep, \
                os.path.dirname(os.path.realpath('__file__')), \
                'valid_reference_fasta_files')
        BWA_INDEX_DIR = os.path.join(os.path.sep, \
                os.path.dirname(os.path.realpath('__file__')), \
                'bwa_indexes')
        LAST_INDEX_DIR = os.path.join(os.path.sep, \
                os.path.dirname(os.path.realpath('__file__')), \
                'last_indexes')

                # # windows version

    if OPER is 'windows':
        config_file = os.path.join(os.path.sep, sys.prefix,
                                   'minup_windows.config')
        logfolder = os.path.join(os.path.sep, sys.prefix,
                                 'minup_run_logs')
        VALID_REF_DIR = os.path.join(os.path.sep, sys.prefix, \
                'valid_reference_fasta_files')
        BWA_INDEX_DIR = os.path.join(os.path.sep, sys.prefix, \
                'bwa_indexes')
        LAST_INDEX_DIR = os.path.join(os.path.sep, sys.prefix, \
                'last_indexes')

#-------------------------------------------------------------------------------

# Parser for command line arguments ...

    parser = \
configargparse.ArgParser(description='minup: A program to analyse minION fast5 files in real-time or post-run.' \
                            , default_config_files=[config_file])
    parser.add( 
        '-w', 
        '--watch-dir',
        type=str,
        required=True,
        default=None,
        help='The path to the folder containing the downloads directory with fast5 reads to analyse - e.g. C:\\data\\minion\\downloads (for windows).',
        dest='watchdir'
    )

    parser.add( 
        '-downloads', 
        '--downloads-dir',
        type=str,
        required=False,
        default='downloads',
        help='The path to the folder containing the downloads directory with fast5 reads to analyse - e.g. C:\\data\\minion\\downloads (for windows).',
        dest='downloads'
    )

    parser.add( 
        '-uploaded', 
        '--uploaded-dir',
        type=str,
        required=False,
        default='uploaded',
        help='The path to the folder containing the uploaded directory with fast5 reads to analyse - e.g. C:\\data\\minion\\downloads (for windows).',
        dest='uploaded'
    )


    parser.add(
        '-dbh',
        '--mysql-host',
        type=str,
        dest='dbhost',
        required=False,
        default='localhost',
        help="The location of the MySQL database. default is 'localhost'.",
    )

    parser.add(
        '-dbu',
        '--mysql-username',
        type=str,
        dest='dbusername',
        required=False,
        default=None,
        help='The MySQL username with create & write privileges on MinoTour.',
    )

    parser.add(
        '-dbp',
        '--mysql-port',
        type=int,
        dest='dbport',
        required=False,
        default=3306,
        help="The MySQL port number, else the default port '3306' is used.",
    )

    parser.add(
        '-pw',
        '--mysql-password',
        type=str,
        dest='dbpass',
        required=True,
        default=None,
        help='The password for the MySQL username with permission to upload to MinoTour.',
    )

    parser.add(
        '-f',
        '--align-ref-fasta',
        type=str,
        required=False,
        default=False,
        help='The reference fasta file to align reads against. Using this option enables read alignment provided LastAl and LastDB are in the path. Leaving this entry blank will upload the data without any alignment. To use multiple reference fasta files input them as one text string seperated by commas (no white spaces) ',
        dest='ref_fasta',
    )

    parser.add(
        '-b',
        '--align-batch-fasta',
        action='store_true',
        required=False,
        default=False,
        help='Align reads in batch processing mode. Assumes the watch-dir (-w) is pointed at a directory with one or more "downloads" folders below it somewhere. Each "downloads" folder can have a subfolder named "reference" containing the fasta file(s) to align the fast5 reads in the corresponding "downloads" folder to ',
        dest='batch_fasta',
    )
    parser.add(
        '-procs',
        '--proc_num',
        type=int,
        dest='procs',
        required=False,
        default=1,
        help='The number of processors to run this on.',
    )

                # parser.add('-n', '--aligning-threads', type=str, required=False, help="The number of threads to use for aligning", default=3, dest='threads')

    parser.add(
        '-u',
        '--minotour-username',
        type=str,
        required=True,
        help='The MinoTour username with permissions to upload data.',
        default=False,
        dest='minotourusername',
    )
    parser.add(
        '-s',
        '--minotour-sharing-usernames',
        type=str,
        required=False,
        default=False,
        help='A comma seperated list (with no whitespaces) of other MinoTour users who will also be able to view the data.',
        dest='view_users',
    )
    parser.add(
        '-o',
        '--flowcell-owner',
        type=str,
        required=False,
        default='minionowner',
        help="The name of the minion owner. 'minionowner' is the default",
        dest='flowcell_owner',
    )
    parser.add(
        '-r',
        '--run-number',
        type=str,
        required=False,
        default=0,
        help='The run number of the flowcell. The default value is 0. ',
        dest='run_num',
    )
    parser.add(
        '-c',
        '--commment-true',
        action='store_true',
        help='Add a comment to the comments field for this run. Follow the prompt once minup starts . ',
        default=False,
        dest='add_comment',
    )
    parser.add(
        '-last',
        '--last-align-true',
        action='store_true',
        help='align reads with LAST',
        default=False,
        dest='last_align',
    )
    parser.add(
        '-bwa',
        '--bwa-align-true',
        action='store_true',
        help='align reads with BWA',
        default=False,
        dest='bwa_align',
    )
    parser.add(
        '-bwa-opts',
        '--bwa-align-options',
        type=str,
        required=False,
        help="BWA options: Enter a comma-seperated list of BWA options without spaces or '-' characters e.g. k12,T0",
        default='T0',
        dest='bwa_options',
    )
    parser.add(
        '-last-opts',
        '--last-align-options',
        type=str,
        required=False,
        help="LAST options: Enter a comma-seperated list of LAST options without spaces or '-' characters e.g. s2,T0,Q0,a1",
        default='s2,T0,Q0,a1',
        dest='last_options',
    )
    parser.add(
        '-pin',
        '--security-pin',
        type=str,
        required=False,
        default=False,
        help='pin number for remote control',
        dest='pin',
    )
    parser.add(
        '-ip',
        '--ip-address',
        type=str,
        required=False,
        default=False,
        help="Used for remote control with option '-pin'. Provide IP address of the computer running minKNOW. The default is the IP address of this computer",
        dest='ip_address',
    )
    parser.add(
        '-t',
        '--insert-tel-true',
        action='store_true',
        #'help='Store all the telemetry data from the read files online. This feature is currently in development.',
        help="DEPRECATED",
        default=False,
        dest='telem',
    )
    parser.add(
        '-pre',
        '--insert-pre-true',
        action='store_true',
        help='Process raw reads prior to basecalling by metrichor.',
        default=False,
        dest='preproc',
    )
    parser.add(
        '-prealign',
        '--align-pre-true',
        action='store_true',
        help='Align raw reads prior to basecalling by metrichor.',
        default=False,
        dest='prealign',
    )
    parser.add(
        '-d',
        '--drop-db-true',
        action='store_true',
        help='Drop existing database if it already exists.',
        default=False,
        dest='drop_db',
    )
    parser.add(
        '-v',
        '--verbose-true',
        help='Print detailed messages while processing files.',
        type=str,
        required=False,
        default='none',
        dest='verbose',
    )
    parser.add(
        '-name',
        '--name-custom',
        type=str,
        required=False,
        default='',
        help='Provide a modifier to the database name. This allows you to upload the same dataset to minoTour more than once. The additional string should be as short as possible.',
        dest='custom_name',
    )
    parser.add(  # MS
        '-cs',
        '--commment-string',
        nargs='+',
        type=str,
        dest='added_comment',
        help='Add given string to the comments field for this run',
        default='',
    )

#                parser.add('-res', '--resume-upload', action='store_true', help="Add files to a partially uploaded database", default=False, dest='resume')

    parser.add(
        '-largerRef',
        '--larger-reference',
        action='store_true',
        #help='Force processng on large reference genomes.',
        help='DEPRECATED',
        default=False,
        dest='largerRef',
    )


    parser.add(
        '-customup',
        '--custom-upload',
        action='store_true',
        help='Stop minUP when cached has remained zero for more than 30 seconds.',
        default=False,
        dest='customup',
    )

    parser.add( # MS ...
        '-qScale',
        '--scale-query',
        action='store_true',
        #help='Scale query by max value in ref.',
        help='DEPRECATED',
        default=False,
        dest='qScale',
    )

    parser.add( # MS ...
        '-qryStartEnd',
        '--prealign-query-start-end-only',
        action='store_true',
        #help='Pre-align using only a window at start and end of the query squiggle? False: pre-align using the whole query squiggle.',
        help="DEPRECATED",
        default=False,
        dest='qryStartEnd',
    )

    parser.add( # MS ...
        '-useHdfTimes',
        '--use-hdf5_file-timestamps',
        action='store_true',
        help='Sort read files using hdt file timestamps rather than file modified times.',
        default=False,
        dest='hdfTimes',
    )

    parser.add( # MS ...
        '-indexToRefDir',
        '--store-indexes-in-ref-folder',
        action='store_true',
        help='Store the bwa/glast index files in the reference genome folder',
        default=False,
        dest='indexToRefDir',
    )

    parser.add( # MS ...
        '-debug',
        '--disable-MyHandler-try-except',
        action='store_true',
        help='Disable MyHandler try excep',
        default=False,
        dest='debug',
    )

    parser.add( # MS ...
        '-standalone',
        '--standalone-mode',
        action='store_true',
        help="DEPRECATED",
        default=False,
        dest='standalone',
    )
    
    parser.add( # MS ...
        '-abs',
        '--aligner_block_size',
        type=int,
        dest='alignerBlockSize',
        required=False,
        default=100,
        help="Number of reads to align on each call to the aligner",
    )
    


    # parser.add('-ver', '--version', action='store_true', help="Report the current version of minUP.", default=False, dest='version') # ML

    # MS ...
    parser.add_argument(
        '-ver',
        '--version',
        action='version',
        version=('%(prog)s version={version}').format(version=__version__),
    )

#-------------------------------------------------------------------------------

    args = parser.parse_args()

    # Set to mySql userrname to minotour username -- rquested by ML 17.1.17...
    args.dbusername = args.minotourusername

    # MS 20.1.17 to handle user defined downloads locations
    args.downloads = args.watchdir

    if args.verbose == 'high':
        print args
        sys.stdout.flush()

    # MS ...
    if args.indexToRefDir is True:
        REF_DIR = os.path.sep.join(args.ref_fasta.split(os.path.sep)[:-1])
        BWA_INDEX_DIR = REF_DIR + '/gbwa.indexes/'
        LAST_INDEX_DIR = REF_DIR + '/glast.indexes/'




    if not os.path.exists(logfolder):
        os.makedirs(logfolder)
    if not os.path.exists(VALID_REF_DIR):
        os.makedirs(VALID_REF_DIR)
    if not os.path.exists(BWA_INDEX_DIR):
        os.makedirs(BWA_INDEX_DIR)
    if not os.path.exists(LAST_INDEX_DIR):
        os.makedirs(LAST_INDEX_DIR)

    # Check inputs are OK...
    if not okSQLname(args.custom_name): # MS
        print 'Error: Invalid SQL name characters in custom_name : %s\nExiting ...' % (args.custom_name)
        sys.stdout.flush()
        sys.exit(1)




    global dbcheckhash
    dbcheckhash = dict()
    dbcheckhash['dbname'] = dict()
    dbcheckhash['barcoded'] = dict()
    dbcheckhash['barcode_info'] = dict()
    dbcheckhash['runindex'] = dict()
    dbcheckhash['modelcheck'] = dict()
    dbcheckhash['logfile'] = dict()

                # dbcheckhash["mafoutdict"]=dict()
                # dbcheckhash["samoutdict"]=dict()

    global ref_fasta_hash
    ref_fasta_hash = dict()

    global dbname
    dbname = str()

    global comments
    comments = dict()

    global connection_pool
    connection_pool = dict()
    try:
        db = MySQLdb.connect(host=args.dbhost, user=args.dbusername,
                             passwd=args.dbpass, port=args.dbport)
        cursor = db.cursor()
    except Exception, err:

        print >> sys.stderr, "Can't connect to MySQL: %s" % err
        sys.stdout.flush()
        sys.exit(1)

                # if args.version == True: # ML
                #                print "minUP version is "+MINUP_VERSION # ML
                #                sys.exit() # ML




    if args.ref_fasta is not False and args.batch_fasta is not False:
        print 'Both --align-ref-fasta (-f) and --align-batch-fasta (-b) were set. Select only one and try again.'
        sys.stdout.flush()
        sys.exit(1)

    if args.last_align is not False and args.bwa_align is not False:
        print 'Both --last-align-true (-last) and --bwa-align-true (-bwa) were set. Select only one and try again.'
        sys.stdout.flush()
        sys.exit(1)

    if args.bwa_align is True and args.ref_fasta is not False:
        process_ref_fasta(args, VALID_REF_DIR \
           , BWA_INDEX_DIR, LAST_INDEX_DIR  \
                       , args.ref_fasta, ref_fasta_hash)

    global bwaclassrunner
    bwaclassrunner = None
    if args.bwa_align is not False:
        from bwaClass import BwaClass
        bwaclassrunner = BwaClass(args.alignerBlockSize, 10)

    comments['default'] = 'No Comment'
    if args.added_comment is not '':  # MS
        comments['default'] = ' '.join(args.added_comment)  # MS
    if args.add_comment is True:
        comment = \
            raw_input('Type comment then press Enter to continue : ')
        comments['default'] = comment


    # MS -- Then, only import this if all is OK ....
    from MyHandler import MyHandler

    print 'monitor started.'
    try:
        signal.signal(signal.SIGINT, emergencyexit2)
        check_read_args = connection_pool, MINUP_VERSION, \
        comments, ref_fasta_hash, dbcheckhash, \
        logfolder, cursor
        observer = Observer()
        event_handler = MyHandler(dbcheckhash, OPER, db, args, xml_file_dict, check_read_args, MINUP_VERSION, bwaclassrunner)
        observer.schedule(event_handler, path=args.watchdir, recursive=True)
        observer.start()
        while True:
            time.sleep(1)

    except (KeyboardInterrupt, SystemExit):
        emergencyexit()
        #print 'stopping monitor....'
        #observer.stop()
        #terminate_minup(args, dbcheckhash, OPER, MINUP_VERSION)

    print "finished."
    observer.join()
    sys.stdout.flush()
    sys.exit(1)



